katdal/visdatav4.py

```diff
@@ -0,0 +1,391 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Data accessor class for data and metadata from various sources in v4 format."""
+
+import logging
+
+import numpy as np
+import katpoint
+
+from .dataset import (DataSet, BrokenFile, Subarray, SpectralWindow,
+                      DEFAULT_SENSOR_PROPS, DEFAULT_VIRTUAL_SENSORS,
+                      _robust_target)
+from .sensordata import SensorCache
+from .categorical import CategoricalData
+from .lazy_indexer import LazyIndexer, LazyTransform
+
+
+logger = logging.getLogger(__name__)
+
+# Simplify the scan activities to derive the basic state of the antenna
+# (slewing, scanning, tracking, stopped)
+SIMPLIFY_STATE = {'scan_ready': 'slew', 'scan': 'scan', 'scan_complete': 'scan',
+                  'load_scan': 'scan', 'track': 'track', 'slew': 'slew'}
+
+SENSOR_PROPS = dict(DEFAULT_SENSOR_PROPS)
+SENSOR_PROPS.update({
+    '*activity': {'greedy_values': ('slew', 'stop'), 'initial_value': 'slew',
+                  'transform': lambda act: SIMPLIFY_STATE.get(act, 'stop')},
+    '*ap_indexer_position': {'initial_value': ''},
+    '*ku_frequency': {'categorical': True},
+    '*noise_diode': {'categorical': True, 'greedy_values': (True,),
+                     'initial_value': 0.0, 'transform': lambda x: x > 0.0},
+    '*serial_number': {'initial_value': 0},
+    '*target': {'initial_value': '', 'transform': _robust_target},
+})
+
+SENSOR_ALIASES = {
+    'nd_coupler': 'dig_noise_diode',
+}
+
+
+def _calc_azel(cache, name, ant):
+    """Calculate virtual (az, el) sensors from actual ones in sensor cache."""
+    real_sensor = 'Antennas/%s/pos_actual_scan_%s' % \
+                  (ant, 'azim' if name.endswith('az') else 'elev')
+    cache[name] = sensor_data = katpoint.deg2rad(cache.get(real_sensor))
+    return sensor_data
+
+
+VIRTUAL_SENSORS = dict(DEFAULT_VIRTUAL_SENSORS)
+VIRTUAL_SENSORS.update({'Antennas/{ant}/az': _calc_azel,
+                        'Antennas/{ant}/el': _calc_azel})
+
+# -----------------------------------------------------------------------------
+# -- CLASS :  VisibilityDataV4
+# -----------------------------------------------------------------------------
+
+
+class VisibilityDataV4(DataSet):
+    """Access format version 4 visibility data and metadata.
+
+    For more information on attributes, see the :class:`DataSet` docstring.
+
+    Parameters
+    ----------
+    metadata : :class:`AttrsSensors` object
+        Metadata attributes and sensors
+    timestamps : array-like of float, length *T*
+        Timestamps at centroids of visibilities in UTC seconds since Unix epoch
+    data : None or :class:`VisFlagsWeights` object, optional
```

Comment:
"Optional" is numpydoc-speak for a parameter having a default value. Since the default value is None, it means that `data` can also be something other than a `VisFlagsWeights` object. You can have the latter situation without also having a default value.

Reply:
Right, but if the default is `None`, it's not normally explicitly listed in the type ([a randomly chosen example](https://docs.scipy.org/doc/numpy/reference/generated/numpy.take.html) and [another](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html)).

Reply:
Point taken. I don't like the extra "None or" either.